File: backend/app/utils/telefone.py
import re


DDDS_VALIDOS = {
    11, 12, 13, 14, 15, 16, 17, 18, 19,
    21, 22, 24, 27, 28,
    31, 32, 33, 34, 35, 37, 38,
    41, 42, 43, 44, 45, 46,
    47, 48, 49,
    51, 53, 54, 55,
    61, 62, 63, 64, 65, 66, 67,
    68, 69,
    71, 73, 74, 75, 77, 79,
    81, 82, 83, 84, 85, 86, 87, 88, 89,
    91, 92, 93, 94, 95, 96, 97, 98, 99,
}


def normalizar_telefone(telefone: str) -> str:
    if not isinstance(telefone, str):
        return ""

    return re.sub(r"\D", "", telefone)


def validar_telefone(telefone: str) -> bool:
    if not isinstance(telefone, str):
        return False

    telefone = telefone.strip()

    # Rejeita letras e caracteres estranhos.
    # Permite apenas números e caracteres de formatação.
    if not re.fullmatch(r"[\d\s().+-]+", telefone):
        return False

    telefone = normalizar_telefone(telefone)

    if len(telefone) not in (10, 11):
        return False

    ddd = int(telefone[:2])

    if ddd not in DDDS_VALIDOS:
        return False

    numero = telefone[2:]

    # Celular
    if len(numero) == 9:
        if not numero.startswith("9"):
            return False

        if len(set(numero)) == 1:
            return False

        return True

    # Telefone fixo
    if len(numero) == 8:
        if numero[0] not in "2345":
            return False

        if len(set(numero)) == 1:
            return False

        return True

    return False


def formatar_telefone(telefone: str) -> str:
    telefone = normalizar_telefone(telefone)

    if len(telefone) == 11:
        return f"({telefone[:2]}) {telefone[2:7]}-{telefone[7:]}"

    if len(telefone) == 10:
        return f"({telefone[:2]}) {telefone[2:6]}-{telefone[6:]}"

    return telefone

File: frontend/src/utils/telefone.ts
const DDDS_VALIDOS = new Set([
  11, 12, 13, 14, 15, 16, 17, 18, 19,
  21, 22, 24, 27, 28,
  31, 32, 33, 34, 35, 37, 38,
  41, 42, 43, 44, 45, 46, 47, 48, 49,
  51, 53, 54, 55,
  61, 62, 63, 64, 65, 66, 67, 68, 69,
  71, 73, 74, 75, 77, 79,
  81, 82, 83, 84, 85, 86, 87, 88, 89,
  91, 92, 93, 94, 95, 96, 97, 98, 99,
]);

export function normalizarTelefone(valor: string): string {
  return valor.replace(/\D/g, "").slice(0, 11);
}

export function formatarTelefone(valor: string): string {
  const numeros = normalizarTelefone(valor);

  if (numeros.length === 0) {
    return "";
  }

  if (numeros.length <= 2) {
    return `(${numeros}`;
  }

  if (numeros.length <= 6) {
    return `(${numeros.slice(0, 2)}) ${numeros.slice(2)}`;
  }

  if (numeros.length <= 10) {
    return `(${numeros.slice(0, 2)}) ${numeros.slice(2, 6)}-${numeros.slice(6)}`;
  }

  return `(${numeros.slice(0, 2)}) ${numeros.slice(2, 7)}-${numeros.slice(7)}`;
}

export function validarTelefone(valor: string): boolean {
  const numeros = normalizarTelefone(valor);

  // Telefone brasileiro deve possuir 10 ou 11 dígitos.
  if (numeros.length !== 10 && numeros.length !== 11) {
    return false;
  }

  const ddd = Number(numeros.slice(0, 2));

  // Verifica se o DDD é válido.
  if (!DDDS_VALIDOS.has(ddd)) {
    return false;
  }

  const numero = numeros.slice(2);

  // Celular: 9 dígitos e começa com 9.
  if (numero.length === 9) {
    if (!numero.startsWith("9")) {
      return false;
    }
  }

  // Telefone fixo: 8 dígitos e começa entre 2 e 5.
  else if (numero.length === 8) {
    if (!"2345".includes(numero[0])) {
      return false;
    }
  }

  else {
    return false;
  }

  // Impede números totalmente repetidos.
  // Exemplos: 999999999, 22222222.
  if (new Set(numero).size === 1) {
    return false;
  }

  return true;
}